lib/nerves_hub_link/downloader/retry_config.ex

```diff
@@ -1,69 +1,77 @@
+      After this amount of time has elapsed, the download is canceled and the download process will crash.
+      The default value is 24 hours.
+      """,
+      type: :non_neg_integer,
+      default: 86_400_000
+    ],
+    time_between_retries: [
+      doc: """
+      Time (in milliseconds) to wait before attempting to retry a download.
+      """,
+      type: :non_neg_integer,
+      default: 15_000
+    ],
+    worst_case_download_speed: [
+      doc: """
+      Worst case download speed specified in bytes per second.
+      This is used to calculate the "worst case" ("sensible") download timeout and is
+      intended to fail faster than waiting for `max_timeout` to elapse.
+      For reference, LTE Cat M1 modems sometimes top out at 32 kbps (30 kbps for some slack).
+      """,
+      type: :non_neg_integer,
+      default: 30_000
+    ]
   ]
 
-  @typedoc """
-  maximum number of disconnects. After this limit is reached
-  the download will be stopped and will no longer be retried
-  """
-  @type max_disconnects :: non_neg_integer()
+  @moduledoc """
+  Download retry configuration.
 
-  @typedoc """
-  time in milliseconds between chunks of data received
-  that once elapsed will trigger a retry. This event counts
-  towards the `max_disconnects` counter
-  """
-  @type idle_timeout :: non_neg_integer()
+  This module provides configuration for how the `Downloader` process will
+  handle disconnects, errors, and timeouts.
 
-  @typedoc """
-  maximum time in milliseconds that a download can exist for.
-  after this amount of time has elapsed, the download is canceled
-  and the download process will crash
-  """
-  @type max_timeout :: non_neg_integer()
+  ## Options
 
-  @typedoc """
-  time in milliseconds to wait before attempting to retry a download
+    #{NimbleOptions.docs(@definition)}
   """
-  @type time_between_retries :: non_neg_integer()
 
-  @typedoc """
-  worst case download speed specified in bytes per second. This is
-  used to calculate the "worst case" download timeout. it is meant to
-  fail faster than waiting for `max_timeout` to elapse
-  """
-  @type worst_case_download_speed :: non_neg_integer()
+  defstruct Keyword.keys(@definition)
 
   @type t :: %__MODULE__{
-          max_disconnects: max_disconnects(),
-          idle_timeout: idle_timeout(),
-          max_timeout: max_timeout(),
-          time_between_retries: time_between_retries(),
-          worst_case_download_speed: worst_case_download_speed()
+          max_disconnects: non_neg_integer(),
+          idle_timeout: non_neg_integer(),
+          max_timeout: non_neg_integer(),
+          time_between_retries: non_neg_integer(),
+          worst_case_download_speed: non_neg_integer()
         }
+
+  @doc """
+  Validates a proposed configuration, raising on error
+  """
+  @spec validate!(Keyword.t()) :: t()
+  def validate!(opts) do
+    validated = NimbleOptions.validate!(opts, @definition)
```

Comment:
Is there a way to validate and use defaults on invalid? We've lasted this long, but it bothers me that none of these options are evaluated until runtime when trying to download a firmware update. Kind of a circular problem to have a bug downloading an update which you would need to download an update to fix 😬

Reply:
Good catch.

The simple fix for the time being is to use `NimbleOptions.validate/2` and pass in the defaults when there is an `{:error, _}`.

Switching to compile time should be in a different PR, as I don't love that we have multiple modules using `Application.get_env`, so I'd like to take a bigger look at this problem.